File: E-Commerce App Project (Tabbed)/Controllers/AccountViewController.m
//
//  AccountViewController.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 7/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "AccountViewController.h"
#import "AccountOPerations.h"

@interface AccountViewController (){
    AccountOperations *accountOperationsObj;
    NSUserDefaults *defaults;
    NSDictionary *userData;
}

@end

@implementation AccountViewController

@synthesize firstName,lastName,email,phone,country,state,city,postalCode,address,goToLoginButtonOutlet;

@synthesize firstNameOutlet,lastNameOutlet,emailOutlet,phoneOutlet,countryOutlet,stateOutlet,cityOutlet,postalCodeOutlet,addressOutlet,userHistoryOutlet,userLogOutOutlet,editUserAccountOutlet;
@synthesize loggedOutAlert;

- (void)viewDidLoad {
    [super viewDidLoad];
    accountOperationsObj = [[AccountOperations alloc]init];
    defaults = [NSUserDefaults standardUserDefaults];
    if (![defaults boolForKey:@"SeesionUserLoggedIN"]) {
        
        firstName.text=@"";
        lastName.text=@"";
        email.text=@"";
        phone.text=@"";
        country.text=@"";
        state.text=@"";
        city.text=@"";
        postalCode.text=@"";
        address.text=@"";
        firstName.hidden=YES;
        lastName.hidden=YES;
        email.hidden=YES;
        phone.hidden=YES;
        country.hidden=YES;
        state.hidden=YES;
        city.hidden=YES;
        postalCode.hidden=YES;
        address.hidden=YES;
        firstNameOutlet.hidden=YES;
        lastNameOutlet.hidden=YES;
        emailOutlet.hidden=YES;
        phoneOutlet.hidden=YES;
        countryOutlet.hidden=YES;
        stateOutlet.hidden=YES;
        cityOutlet.hidden=YES;
        postalCodeOutlet.hidden=YES;
        addressOutlet.hidden=YES;
        goToLoginButtonOutlet.hidden=NO;
        userHistoryOutlet.enabled=NO;
        userHistoryOutlet.tintColor=[UIColor clearColor];
        userLogOutOutlet.enabled=NO;
        userLogOutOutlet.tintColor=[UIColor clearColor];
//        editUserAccountOutlet.enabled=NO;
//        editUserAccountOutlet.tintColor=[UIColor clearColor];
    }else{
        firstName.hidden=NO;
        lastName.hidden=NO;
        email.hidden=NO;
        phone.hidden=NO;
        country.hidden=NO;
        state.hidden=NO;
        city.hidden=NO;
        postalCode.hidden=NO;
        address.hidden=NO;
        firstNameOutlet.hidden=NO;
        lastNameOutlet.hidden=NO;
        emailOutlet.hidden=NO;
        phoneOutlet.hidden=NO;
        countryOutlet.hidden=NO;
        stateOutlet.hidden=NO;
        cityOutlet.hidden=NO;
        postalCodeOutlet.hidden=NO;
        addressOutlet.hidden=NO;
        goToLoginButtonOutlet.hidden=YES;
        userHistoryOutlet.enabled=YES;
        userHistoryOutlet.tintColor=[UIColor colorNamed:@"Cornflower Blue"];
        userLogOutOutlet.enabled=YES;
        userLogOutOutlet.tintColor=[UIColor colorNamed:@"Cornflower Blue"];
//        editUserAccountOutlet.enabled=YES;
//        editUserAccountOutlet.tintColor=[UIColor colorNamed:@"Cornflower Blue"];
        
        userData = [defaults dictionaryForKey:@"LoggedInUsersDetail"];
        
        firstName.text=[userData valueForKey:@"firstName"];
        lastName.text=[userData valueForKey:@"lastName"];
        email.text=[userData valueForKey:@"usersEmail"];
        phone.text=[userData valueForKey:@"phone"];
        country.text=[userData valueForKey:@"country"];
        state.text=[userData valueForKey:@"state"];
        city.text=[userData valueForKey:@"city"];
        postalCode.text=[userData valueForKey:@"postalCode"];
        address.text=[userData valueForKey:@"address"];
    }
    
    loggedOutAlert = [UIAlertController
                             alertControllerWithTitle:@"Log Out?"
                             message:@"Are you sure?"
                             preferredStyle:UIAlertControllerStyleAlert];
    
    // Do any additional setup after loading the view.
    UIAlertAction* noButton = [UIAlertAction
                                actionWithTitle:@"Yes, Log Out!"
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    //Handle your yes please button action here
                                    [self->defaults setBool:NO forKey:@"SeesionUserLoggedIN"];
                                    [self->defaults setObject:@"" forKey:@"SessionLoggedInuserEmail"];
                                    [self->defaults setObject:@"" forKey:@"LoggedInUsersDetail"];
                                    [defaults synchronize];
                                }];
    UIAlertAction* yesButton = [UIAlertAction
                               actionWithTitle:@"Log Back In!"
                               style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction * action) {
                                   
                               }];
    
    [loggedOutAlert addAction:yesButton];
    [loggedOutAlert addAction:noButton];
}
- (void)viewDidAppear:(BOOL)animated{
    [self.view reloadInputViews];
    if (![defaults boolForKey:@"SeesionUserLoggedIN"]) {
        firstName.text=@"";
        lastName.text=@"";
        email.text=@"";
        phone.text=@"";
        country.text=@"";
        state.text=@"";
        city.text=@"";
        postalCode.text=@"";
        address.text=@"";
        firstName.hidden=YES;
        lastName.hidden=YES;
        email.hidden=YES;
        phone.hidden=YES;
        country.hidden=YES;
        state.hidden=YES;
        city.hidden=YES;
        postalCode.hidden=YES;
        address.hidden=YES;
        firstNameOutlet.hidden=YES;
        lastNameOutlet.hidden=YES;
        emailOutlet.hidden=YES;
        phoneOutlet.hidden=YES;
        countryOutlet.hidden=YES;
        stateOutlet.hidden=YES;
        cityOutlet.hidden=YES;
        postalCodeOutlet.hidden=YES;
        addressOutlet.hidden=YES;
        goToLoginButtonOutlet.hidden=NO;
        userHistoryOutlet.enabled=NO;
        userHistoryOutlet.tintColor=[UIColor clearColor];
        userLogOutOutlet.enabled=NO;
        userLogOutOutlet.tintColor=[UIColor clearColor];
//        editUserAccountOutlet.enabled=NO;
//        editUserAccountOutlet.tintColor=[UIColor clearColor];
    }else{
        firstName.hidden=NO;
        lastName.hidden=NO;
        email.hidden=NO;
        phone.hidden=NO;
        country.hidden=NO;
        state.hidden=NO;
        city.hidden=NO;
        postalCode.hidden=NO;
        address.hidden=NO;
        firstNameOutlet.hidden=NO;
        lastNameOutlet.hidden=NO;
        emailOutlet.hidden=NO;
        phoneOutlet.hidden=NO;
        countryOutlet.hidden=NO;
        stateOutlet.hidden=NO;
        cityOutlet.hidden=NO;
        postalCodeOutlet.hidden=NO;
        addressOutlet.hidden=NO;
        goToLoginButtonOutlet.hidden=YES;
        userHistoryOutlet.enabled=NO;
        userHistoryOutlet.tintColor=[UIColor clearColor];
        userLogOutOutlet.enabled=YES;
        userLogOutOutlet.tintColor=[UIColor colorNamed:@"Cornflower Blue"];
//        editUserAccountOutlet.enabled=YES;
//        editUserAccountOutlet.tintColor=[UIColor colorNamed:@"Cornflower Blue"];
        
        userData = [defaults dictionaryForKey:@"LoggedInUsersDetail"];
        
        firstName.text=[userData valueForKey:@"firstName"];
        lastName.text=[userData valueForKey:@"lastName"];
        email.text=[userData valueForKey:@"usersEmail"];
        phone.text=[userData valueForKey:@"phone"];
        country.text=[userData valueForKey:@"country"];
        state.text=[userData valueForKey:@"state"];
        city.text=[userData valueForKey:@"city"];
        postalCode.text=[userData valueForKey:@"postalCode"];
        address.text=[userData valueForKey:@"address"];
    }
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (IBAction)theEditButtonForLoginOrEdit:(id)sender {
    if (![defaults boolForKey:@"SeesionUserLoggedIN"]) {
        [self performSegueWithIdentifier:@"loginAccountSegue" sender:nil];
    }else{
        [self performSegueWithIdentifier:@"editAccountSegue" sender:nil];
    }
}

- (IBAction)goToLoginButton:(id)sender {
    [self performSegueWithIdentifier:@"loginAccountSegue" sender:nil];
}

- (IBAction)userLogoutButton:(id)sender {
    [self presentViewController:loggedOutAlert animated:YES completion:nil];
}
@end

File: E-Commerce App Project (Tabbed)/Controllers/AccountViewController.swift
//
//  AccountViewController.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift with MVVM pattern
//

import UIKit

class AccountViewController: UIViewController {
    
    // MARK: - ViewModel
    private let viewModel = AccountViewModel()
    
    // MARK: - IBOutlets
    @IBOutlet var firstName: UILabel!
    @IBOutlet var lastName: UILabel!
    @IBOutlet var email: UILabel!
    @IBOutlet var phone: UILabel!
    @IBOutlet var country: UILabel!
    @IBOutlet var state: UILabel!
    @IBOutlet var city: UILabel!
    @IBOutlet var postalCode: UILabel!
    @IBOutlet var address: UILabel!
    @IBOutlet var goToLoginButtonOutlet: UIButton!
    
    @IBOutlet var firstNameOutlet: UILabel!
    @IBOutlet var lastNameOutlet: UILabel!
    @IBOutlet var emailOutlet: UILabel!
    @IBOutlet var phoneOutlet: UILabel!
    @IBOutlet var countryOutlet: UILabel!
    @IBOutlet var stateOutlet: UILabel!
    @IBOutlet var cityOutlet: UILabel!
    @IBOutlet var postalCodeOutlet: UILabel!
    @IBOutlet var addressOutlet: UILabel!
    @IBOutlet var userHistoryOutlet: UIBarButtonItem!
    @IBOutlet var userLogOutOutlet: UIBarButtonItem!
    @IBOutlet var editUserAccountOutlet: UIBarButtonItem!
    
    var loggedOutAlert: UIAlertController!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupLogoutAlert()
        bindViewModel()
        viewModel.refreshUserData()
    }
    
    private func setupLogoutAlert() {
        loggedOutAlert = UIAlertController(
            title: "Log Out?",
            message: "Are you sure?",
            preferredStyle: .alert
        )
        
        let logoutAction = UIAlertAction(title: "Yes, Log Out!", style: .default) { [weak self] _ in
            self?.viewModel.logout()
        }
        
        let cancelAction = UIAlertAction(title: "Log Back In!", style: .default, handler: nil)
        
        loggedOutAlert.addAction(cancelAction)
        loggedOutAlert.addAction(logoutAction)
    }
    
    private func bindViewModel() {
        viewModel.isLoggedIn.bind { [weak self] isLoggedIn in
            guard let self = self else { return }
            if isLoggedIn {
                self.showUserInfo()
            } else {
                self.hideUserInfo()
            }
        }
        
        viewModel.userFirstName.bind { [weak self] value in
            self?.firstName.text = value
        }
        
        viewModel.userLastName.bind { [weak self] value in
            self?.lastName.text = value
        }
        
        viewModel.userEmail.bind { [weak self] value in
            self?.email.text = value
        }
        
        viewModel.userPhone.bind { [weak self] value in
            self?.phone.text = value
        }
        
        viewModel.userCountry.bind { [weak self] value in
            self?.country.text = value
        }
        
        viewModel.userState.bind { [weak self] value in
            self?.state.text = value
        }
        
        viewModel.userCity.bind { [weak self] value in
            self?.city.text = value
        }
        
        viewModel.userPostalCode.bind { [weak self] value in
            self?.postalCode.text = value
        }
        
        viewModel.userAddress.bind { [weak self] value in
            self?.address.text = value
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        view.reloadInputViews()
        viewModel.refreshUserData()
    }
    
    private func hideUserInfo() {
        firstName.isHidden = true
        lastName.isHidden = true
        email.isHidden = true
        phone.isHidden = true
        country.isHidden = true
        state.isHidden = true
        city.isHidden = true
        postalCode.isHidden = true
        address.isHidden = true
        firstNameOutlet.isHidden = true
        lastNameOutlet.isHidden = true
        emailOutlet.isHidden = true
        phoneOutlet.isHidden = true
        countryOutlet.isHidden = true
        stateOutlet.isHidden = true
        cityOutlet.isHidden = true
        postalCodeOutlet.isHidden = true
        addressOutlet.isHidden = true
        goToLoginButtonOutlet.isHidden = false
        userHistoryOutlet.isEnabled = false
        userHistoryOutlet.tintColor = .clear
        userLogOutOutlet.isEnabled = false
        userLogOutOutlet.tintColor = .clear
    }
    
    private func showUserInfo() {
        firstName.isHidden = false
        lastName.isHidden = false
        email.isHidden = false
        phone.isHidden = false
        country.isHidden = false
        state.isHidden = false
        city.isHidden = false
        postalCode.isHidden = false
        address.isHidden = false
        firstNameOutlet.isHidden = false
        lastNameOutlet.isHidden = false
        emailOutlet.isHidden = false
        phoneOutlet.isHidden = false
        countryOutlet.isHidden = false
        stateOutlet.isHidden = false
        cityOutlet.isHidden = false
        postalCodeOutlet.isHidden = false
        addressOutlet.isHidden = false
        goToLoginButtonOutlet.isHidden = true
        userHistoryOutlet.isEnabled = false
        userHistoryOutlet.tintColor = .clear
        userLogOutOutlet.isEnabled = true
        userLogOutOutlet.tintColor = UIColor(named: "Cornflower Blue")
    }
    
    @IBAction func theEditButtonForLoginOrEdit(_ sender: Any) {
        if viewModel.isLoggedIn.value {
            performSegue(withIdentifier: "editAccountSegue", sender: nil)
        } else {
            performSegue(withIdentifier: "loginAccountSegue", sender: nil)
        }
    }
    
    @IBAction func goToLoginButton(_ sender: Any) {
        performSegue(withIdentifier: "loginAccountSegue", sender: nil)
    }
    
    @IBAction func userLogoutButton(_ sender: Any) {
        present(loggedOutAlert, animated: true, completion: nil)
    }
}
